tests/gtests/internals/test_sdpa.cpp

```diff
@@ -0,0 +1,1817 @@
+        //                out[i], input[i], zero_points[g], scales[g]);
+        //    }
+        //}
+    }
+    return out;
+}
+
+void transpose(const dnnl::engine &eng, memory &out, memory &in) {
+    dnnl::stream s(eng);
+
+    void *ptr2 = out.map_data();
+    void *ptr1 = in.map_data();
+
+    std::memcpy(ptr2, ptr1, in.get_desc().get_size());
+    in.unmap_data(ptr1);
+    out.unmap_data(ptr2);
+}
+
+void transpose_strides(const dnnl::engine &eng, memory &out, memory &in) {
+    dnnl::stream s(eng);
+
+    if (out.get_desc().get_data_type() == mdt::u4
+            || out.get_desc().get_data_type() == mdt::s4) {
+        auto desc = in.get_desc();
+        auto dims = desc.get_dims();
+        auto strides = desc.get_strides();
+        auto strides_t = out.get_desc().get_strides();
+
+        char *mapped_ptr = (char *)in.map_data();
+        char *mapped_ptr_t = (char *)out.map_data();
+        for_(int l = 0; l < dims[0]; l++)
+        for_(int k = 0; k < dims[1]; k++)
+        for_(int j = 0; j < dims[2]; j++)
+        for (int i = 0; i < dims[3]; i++) {
+            int is_odd = i % 2;
+            int is_odd_t = j % 2;
+
+            auto offset = l * strides[0] + k * strides[1] + j * strides[2]
+                    + i * strides[3];
+            offset /= 2;
+
+            auto offset_t = l * strides_t[0] + k * strides_t[1]
+                    + j * strides_t[2] + i * strides_t[3];
+            offset_t /= 2;
+
+            auto &val = mapped_ptr[offset];
+            auto &val_t = mapped_ptr_t[offset_t];
+
+            char bits;
+            if (is_odd) {
+                bits = val & 0xf0;
+                bits >>= 4;
+            } else {
+                bits = val & 0x0f;
+            }
+            if (is_odd_t) {
+                val_t |= (bits << 4);
+            } else {
+                val_t |= bits;
+            }
+        }
+        in.unmap_data(mapped_ptr);
+        out.unmap_data(mapped_ptr_t);
+    } else {
+        dnnl::reorder(in, out).execute(s, in, out);
+    }
+}
+
+memory::dims double_mb(const memory::dims &dims) {
+    memory::dims ret = dims;
+    ret[0] *= 2;
+    return ret;
+}
+
+memory double_and_resize(const memory::desc &desc, dnnl::engine &eng) {
+    dnnl::stream s(eng);
+    memory::dims dims2 = double_mb(desc.get_dims());
+    auto desc2 = memory::desc(dims2, desc.get_data_type(), desc.get_strides());
+
+    dnnl_memory_t mem2;
+    CHECK(dnnl_memory_create(
+            &mem2, desc2.get(), eng.get(), DNNL_MEMORY_ALLOCATE));
+
+    void *mapped_ptr = nullptr;
+    CHECK(dnnl_memory_map_data(mem2, &mapped_ptr));
+
+    for (size_t i = 0; i < desc2.get_size(); i++) {
+        ((uint8_t *)mapped_ptr)[i] = 0xFF;
+    }
+    CHECK(dnnl_memory_unmap_data(mem2, mapped_ptr));
+
+    void *handle;
+    CHECK(dnnl_memory_get_data_handle(mem2, &handle));
+    return memory(desc, eng, handle);
+}
+
+sdpa_tensors_t get_descriptors(dnnl::engine &eng, sdpa_dims_t p) {
```

Comment:
```suggestion
sdpa_tensors_t get_descriptors(dnnl::engine &eng, const sdpa_dims_t &p) {
```

I'd probably run this code with clang-tidy and see what it reports. Would save extra touches after promotion.